Display account created/linked message on enterprise landing page

zubair-arbi opened the pull request:
ENT-347
@asadiqbal08 @saleem-latif @douglashall 

**Description:**  	Display account created/linked message on enterprise course enrollment page (enterprise landing page).

**JIRA:** [ENT-347](https://openedx.atlassian.net/browse/ENT-347)

**Dependencies:** N/A

**Merge deadline:** 19 May, 2017

**Installation instructions:** Install `edx-enterprise` from this branch `zub/ENT-347-enterprise-landing-page-display-messages` in `edx-platform`.

**Testing instructions:**

1. Go to enterprise customers django admin at `admin/enterprise/enterprisecustomer/`
2. Add enterprise customer with logo and with consent disabled and link it with an `idp`.
3. Create/publish a course in studio
4. As an anonymous user view the enterprise landing page for the enterprise course by accessing the enterprise course enrollment url with enterprise customer uuid from step no 2, e.g: http://zubair.localhost:8000/enterprise/72416e52-8c77-4860-9584-15e5b06220fb/course/course-v1:edX+TEST_01+2016_R1/enroll/
5. The learner will be redirected to `idp` login page related to the enterprise and on successful login if learner will create a new account at edX for linking with enterprise then learner will view both **success** and **info** messages. If the learner links and existing activated account at edX then the learner will only see **success** message on enterprise landing page
6. Repeat from step **2** by enabling consent. The difference in flow will be that now learner will have to accept the consent after successful linking of account with the enterprise.

**Merge checklist:**

- [ ] Regenerate requirements with `make upgrade && make requirements` (and make sure to fix any errors).
  **DO NOT** just add dependencies to `requirements/*.txt` files.
- [x] All reviewers approved
- [x] CI build is green
- [x] Version bumped
- [x] Changelog record added
- [ ] Documentation updated (not only docstrings)
- [x] Commits are (reasonably) squashed
- [ ] Translations are updated
- [ ] PR author is listed in AUTHORS

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPi after tag-triggered build is finished.
- [ ] Delete working branch (if not needed anymore)

**Author concerns:** List any concerns about this PR - inelegant 
solutions, hacks, quick-and-dirty implementations, concerns about 
migrations, etc.

douglashall commented on enterprise/static/enterprise/enterprise_course_enrollment_page.css:
```diff
@@ -207,3 +207,35 @@ body, .zebra-stripe > :nth-child(2n+1), .depth-0 {
+  background-color: #f2dede;
+  border-color: #ebccd1;
+  color: #a94442;
+}
```
Are there variables we could use for the colors here?

zubair-arbi replied:
@douglashall we don't have setup for `scss` compiling so for now we are directly using the colors codes provided by Christine instead of defining variables for these color codes.

douglashall commented on enterprise/views.py:
```diff
@@ -468,6 +478,26 @@ def get_enterprise_course_enrollment_page(self, request, enterprise_uuid, course
+        if not request.user.is_authenticated():
+            return redirect(
+                '{login_url}?next={current_url}?{query_string}'.format(
+                    login_url=reverse('signin_user'),
```
Don't use reverse here. We may someday want edx-enterprise to run as a separate IDA.

douglashall commented on enterprise/views.py:
```diff
@@ -468,6 +478,26 @@ def get_enterprise_course_enrollment_page(self, request, enterprise_uuid, course
+        # Enterprise-linked SSO and the pipeline will get them back here.
+        if not request.user.is_authenticated():
+            return redirect(
+                '{login_url}?next={current_url}?{query_string}'.format(
```
Use `urljoin`.

douglashall commented on enterprise/templates/enterprise/enterprise_course_enrollment_page.html:
```diff
@@ -33,6 +33,30 @@
+                  </div>
+                % endfor
+              </div>
+          % endif
```
This would probably make for a good custom template tag, which brings up a good point. We should probably be using Django templates instead of mako templates.

zubair-arbi replied:
@douglashall Nice suggestion. We need to make sure that open craft is also aware of this move and currently not making changes on this template.
cc: @bradenmacdonald @haikuginger

bradenmacdonald replied:
Yes, we should definitely be using django templates! Thanks for the heads up.

asadiqbal08 commented on enterprise/tpa_pipeline.py:
```diff
@@ -142,11 +143,18 @@ def redirect_to_consent():
         # This enterprise customer attached to this pipeline element does not request data sharing consent;
         # proceed with the creation of a link between the user and the enterprise customer, then exit.
-        ec_user, _ = EnterpriseCustomerUser.objects.get_or_create(
+        ec_user, __ = EnterpriseCustomerUser.objects.get_or_create(
```
nit: please use `enterprise_customer_user` instead of `ec_*`

asadiqbal08 commented on enterprise/tpa_pipeline.py:
```diff
@@ -142,11 +143,18 @@ def redirect_to_consent():
             user_id=user.id
         )
 
+        messages.success(backend.strategy.request, _('Thank you for creating an account with edX.'))
```
`edX` or `platform_name` ?

zubair-arbi replied:
yes. Thanks for pointing this out.

asadiqbal08 commented on enterprise/tpa_pipeline.py:
```diff
@@ -142,11 +143,18 @@ def redirect_to_consent():
+        messages.success(backend.strategy.request, _('Thank you for creating an account with edX.'))
+        if not user.is_active:
+            messages.info(
+                backend.strategy.request,
```
nit: We **have** just sent you **an** email ...  (representing an action that has been done in recently) .

asadiqbal08 commented on enterprise/views.py:
```diff
@@ -391,6 +394,13 @@ def post_account_consent(self, request, consent_provided):
+        if not user.is_active:
+            messages.info(
+                request,
+                _('We just sent you email. Activate your account before you next log in.')
```
nit: We **have** just sent you **an** email ...

zubair-arbi commented:
Here is the sandbox (with `sso` enabled): https://zubair-arbi.sandbox.edx.org/enterprise/c62208d9-ab13-4116-aeb3-0a797d17839e/course/course-v1:HKUST+EC001+2017_R1/enroll/

zubair-arbi commented:
@douglashall @asadiqbal08 I have addressed most of your feedback except conversion to django template (waiting for consensus with open craft) and messages content (waiting for response on Jira ticket). Also fixed and updated tests for coverage of new django messages.

douglashall commented:
@zubair-arbi Can you create a story for converting the course enrollment landing page to a Django template? We can finish our work this sprint and pick that conversion up next sprint. Sound good?

zubair-arbi commented:
@douglashall Sounds good. I will create a Jira ticket to track the conversion work.

zubair-arbi commented:
@douglashall I have created [ENT-399](https://openedx.atlassian.net/browse/ENT-399) to convert the enterprise landing page to django template.

douglashall commented on enterprise/views.py:
```diff
@@ -468,6 +482,24 @@ def get_enterprise_course_enrollment_page(self, request, enterprise_uuid, course
+                current_url=quote(request.get_full_path()),
+                query_string=urlencode({'tpa_hint': enterprise_customer.identity_provider})
+            )
+            return redirect('{login_url}?next={next_url}'.format(login_url='/login', next_url=next_url))
```
Does the `tpa_hint` here prevent us from using the `@login_required` decorator? If so, can we create a new decorator that implements this logic?

zubair-arbi replied:
@douglashall yes. We need `tpa_hint` in the redirect url which depends on enterprise provider and its linked `idp`. I believe we can create a decorator which will implement this logic but in the decorator we will have to query database. Also I am not sure if this decorator will be reusable with other urls as this is the only url with enterprise customer uuid. Please let me know your thoughts.

douglashall replied:
Yes, I think we should implement a decorator called something like `enterprise_login_required` which implements this logic.

zubair-arbi commented:
@asadiqbal08 @douglashall I have added more tests to complete the coverage. Its ready for another review.

douglashall commented on enterprise/views.py:
```diff
@@ -460,6 +481,8 @@ def get_enterprise_course_enrollment_page(self, request, enterprise_uuid, course
         """
         Render enterprise specific course track selection page.
 
+        Un authenticated learners will be redirected to enterprise-linked SSO.
```
"Unauthenticated" should be one word.

zubair-arbi commented:
@douglashall I have added a new decorator "`enterprise_login_required`" and refactored the code a bit. @saleem-latif is working on the decorator tests. Otherwise its ready for another review.

douglashall commented on enterprise/utils.py:
```diff
@@ -527,3 +532,61 @@ def filter_audit_course_modes(enterprise_customer, course_modes):
+
+def enterprise_login_required(view):
+    """
+    Decorator for enterprise views.
```
Can you make this line a short description of what the decorator does for enterprise views?

zubair-arbi commented:
@asadiqbal08 please review.

asadiqbal08 commented:
👍
